experimental/memory.go

```diff
@@ -0,0 +1,31 @@
+package experimental
+
+import (
+	"context"
+
+	"github.com/tetratelabs/wazero/internal/ctxkey"
+)
+
+// MemoryAllocator is a memory allocation hook.
+type MemoryAllocator interface {
```

Comment:
If this weren't an experimental API, I guess currently it would have to go on `ModuleConfig` rather than `RuntimeConfig` since it looks like it can only return a single buffer. An alternative that is runtime-scoped could return an interface

```
type Buffer interface {
  Bytes() []byte
}

type MemoryAllocator interface {
  Make(...) Buffer

  Grow(Buffer, uint64) Buffer

  Free(Buffer)
```

`Memory` would store a reference to both Buffer and `[]byte`, later for all memory operations and former for interacting with the allocator.

This somehow seems like what I would expect from an allocator but it's not clearly better. I think with the current one, I would at least rename it to `ModuleMemoryAllocator` to make it more obvious it is only scoped to a single instance of a module (of course not using multi-memory which I don't think we need to consider, but just as a FYI).

Reply:
I realized another idea that's probably better than above is

```
type Buffer interface {
  Bytes() []byte

  Grow(uint64)

  Free()
}

type MemoryAllocator interface {
  Make(...) Buffer
```

This is only if making allocator a runtime config rather than a module config. Somehow these seem closer to how other APIs like malloc look

Reply:
Thanks! That's an improvement, but I still trying to come up with an API that's safe to use by shared memories. Unfortunately, it's likely not very ergonomic…

I mean, `Grow` needs to update the `[]byte` inside `MemoryInstance` atomically. The only way I see to do that, is to pass a `*[]byte` into `Grow`?

Reply:
Sorry for the churn. New attempt.

Also fixes a race (I think) in existing shared memory, where the size of shared memory could go back (become smaller).

Reply:
Previously we had a lock around Grow to make it atomic but we removed it since it didn't seem to make sense if the guest doesn't itself lock around Grow, and it's at least the case with wasi-libc. So not sure if we need to be worried here either, or is it a bad assumption?

Reply:
Tbh, I don't know.
But the previous atomic write doesn't seem to be guaranteeing anything much?
I mean, yes, it makes the write visible to other threads, but it doesn't ensure length can't go down.
I don't know what wasi-libc is doing if it needs locks when multiple threads try to grow concurrently.
Do you have a code pointer?

Reply:
I think the code pointer is this one

https://github.com/WebAssembly/wasi-libc/blob/main/dlmalloc/src/malloc.c#L262

Every malloc call is surrounded by a lock. Since only malloc can grow, this implicitly locks grow. This seems to make sense to me as grow looks closer to sbrk (not allowed to be called concurrently AFAIK), then mmap. I'm not sure how to reason about the return value of grow if it could be called concurrently,

> but it doesn't ensure length can't go down.

Currently the code is assuming non-concurrent grow so only cares about the visibility and the length wouldn't go down (I think, let me know if I missed something different).

I filed https://github.com/WebAssembly/threads/issues/224 to see what the spec may have intended. I would personally still assume grow is called non-concurrently until hearing on it since I think it's been our assumption since https://github.com/tetratelabs/wazero/pull/1755

Reply:
So what does it mean that we care about visibility? Who/what views it?

If it's locked in the guest, which makes sense, that mutex's release takes care to publish the changes. We wouldn't need to do anything about it, right?

Reply:
I'm asking because I think the allocator needs the same treatment, and I'd like to understand what (if anything) I should document.

From what you're saying, custom the allocators for shared memories wouldn't need a mutex, and ISTM they don't need this atomic either.

Reply:
Hmm tried looking for a conversation I thought we had about visibility but can't find it maybe it was an unmerged PR.

Basically, the mutex is around grow but memory length is used in bounds check and is read without the mutex. So it seems theoretically possible to malloc a pointer in a thread that called grow and read in a different one where bounds check fails.

Since this is only about the visibility of size to bounds checks, I believe this doesn't affect custom allocators as that part is managed by wazero.

Reply:
OK, that makes sense. Will revert to the simpler version.